### PR TITLE
Change QSA patch to use native QSA and filter the results

### PR DESCRIPTION
**Pull request by muan**
In using Web Components polyfill we found that the patched `querySelectorAll` has a serious performance impact on our pages, especially when running on long pages with 100+ selectors. After looking into the patched implementation, it became clear why. However in investigating, I also found this comment suggesting the filtered QSA approach:

https://github.com/muan/shadydom/blob/2e4b73247a2bf413148acb2970f2a0dea886c837/src/patch-builtins.js#L122

So, as an experiment I tried it out to see if it can actually make performance better, and it did. It's possible that I might have missed some Shady DOM behaviors, but I hope by basing the filter on existing methods I have avoided as much of that as possible. And this also still uses the original qsa patch for all shady elements to minimize the scope. 

It seems that calling `flush()` probably [should be avoided](https://github.com/webcomponents/shadydom#limitations), however I didn't figure out a way to do that. 😔

Over on https://github.com/muan/shadydom/compare/qsa...muan:qsa-html?expand=1 I have created a test page to see if the results are the same as the native implementation, and they are matching. 👌🏻

Here are the performance test pages for `document.querySelectorAll(selectors)` (check console):

**No polyfill**: http://muan.github.io/playground/native.html
**Current patch**: http://muan.github.io/playground/shady.html
**This patch**: http://muan.github.io/playground/patched.html

Results in Chrome:

The tests are all passing locally.

cc @josh @dgraham

### CODE REVIEW

File: src/patch-builtins.js [hunk 1]
```diff
@@ -137,9 +137,25 @@ let fragmentMixin = {
...
-    });
+    var that =  this
+    if (this instanceof Document) {
+      flush();
+      let result = nativeDQSA.call(this, selector);
+      result = Array.prototype.filter.call(result, function (n) {
+        return !utils.ownerShadyRootForNode(n);
+      });
+      return result;
+    } else if (!this.__shady && this instanceof Element) {
```
Comment (TimvdLippe) [inline src/patch-builtins.js]: This codepath is puzzling to me. Could you add clarifying comments to explain the condition and why the filter is applied to childNodes later?

E.g. why not loop through the childNodes and check for matchesselector? That would prevent looping through all other deep children.

**Comment by notwaldorf** [comment]
cc @azakus

**Comment by sorvell** [comment]
Thanks for submitting this. We've also considered this approach, but we've been hesitant because it misses undistributed elements since these are not actually in the DOM (It's not enough to check if the element itself has a `shadowRoot` as this PR does.).

We've considered living with this limitation because of the major performance improvement, but we're also consider perhaps leaving undistributed elements in the dom, hidden to avoid this issue. We want to experiment with that approach before deciding whether to merge this.

**Comment by sorvell** [comment]
As a workaround in the meantime, you can use the captured native element `querySelectorAll` something like this (you can call this on any element, but not on document).

```
function fasterQuerySelectorAll(element, selector) {
  const nodes = Array.from(ShadyDOM.nativeMethods.querySelectorAll.call(element, selector));
  return nodes.filter((e) => e.getRootNode() == element.getRootNode());
}
```

**Comment by TimvdLippe** [comment]
@muan Did #216 supersede this PR?

**Comment by muan** [comment]
> @muan Did #216 supersede this PR?

I think so. I'll give that a go. Thanks all!

**Pull request closed without merging**